aws/resource_aws_cloudhsm2_hsm.go

```diff
@@ -0,0 +1,285 @@
+		}
+
+		return hsm, *hsm.State, err
+	}
+}
+
+func resourceAwsCloudHsm2HsmCreate(d *schema.ResourceData, meta interface{}) error {
+	cloudhsm2 := meta.(*AWSClient).cloudhsmv2conn
+
+	clusterId := d.Get("cluster_id").(string)
+
+	cluster, err := describeCloudHsm2Cluster(clusterId, meta)
+
+	if cluster == nil {
+		log.Printf("[WARN] Error on retrieving CloudHSMv2 Cluster: %s %s", clusterId, err)
+		return err
+	}
+
+	availabilityZone := d.Get("availability_zone").(string)
+	if len(availabilityZone) == 0 {
+		subnetId := d.Get("subnet_id").(string)
+		for az, sn := range cluster.SubnetMapping {
+			if *sn == subnetId {
+				availabilityZone = az
+			}
+		}
+	}
+
+	input := &cloudhsmv2.CreateHsmInput{
+		ClusterId:        aws.String(clusterId),
+		AvailabilityZone: aws.String(availabilityZone),
+	}
+
+	ipAddress := d.Get("ip_address").(string)
+	if len(ipAddress) != 0 {
+		input.IpAddress = aws.String(ipAddress)
+	}
+
+	log.Printf("[DEBUG] CloudHSMv2 HSM create %s", input)
+
+	var output *cloudhsmv2.CreateHsmOutput
+
+	errRetry := resource.Retry(180*time.Second, func() *resource.RetryError {
+		var err error
+		output, err = cloudhsm2.CreateHsm(input)
+		if err != nil {
+			if isAWSErr(err, cloudhsmv2.ErrCodeCloudHsmInternalFailureException, "request was rejected because of an AWS CloudHSM internal failure") {
+				log.Printf("[DEBUG] CloudHSMv2 HSM re-try creating %s", input)
+				return resource.RetryableError(err)
+			}
+			return resource.NonRetryableError(err)
+		}
+		return nil
+	})
+
+	if errRetry != nil {
+		return errRetry
+	}
+
+	d.SetId(aws.StringValue(output.Hsm.HsmId))
+	log.Printf("[INFO] CloudHSMv2 HSM Id: %s", d.Id())
+	log.Println("[INFO] Waiting for CloudHSMv2 HSM to be available")
+
+	stateConf := &resource.StateChangeConf{
+		Pending:    []string{cloudhsmv2.HsmStateCreateInProgress, "destroyed"},
+		Target:     []string{cloudhsmv2.HsmStateActive},
+		Refresh:    resourceAwsCloudHsm2HsmRefreshFunc(d, meta),
+		Timeout:    d.Timeout(schema.TimeoutCreate),
+		MinTimeout: 30 * time.Second,
+		Delay:      30 * time.Second,
+	}
+
+	// Wait, catching any errors
+	_, errWait := stateConf.WaitForState()
+	if errWait != nil {
+		return fmt.Errorf("[WARN] Error waiting for CloudHSMv2 HSM state to be \"ACTIVE\": %s", errWait)
+	}
+
+	return resourceAwsCloudHsm2HsmRead(d, meta)
+}
+
+func resourceAwsCloudHsm2HsmRead(d *schema.ResourceData, meta interface{}) error {
+
+	hsm, err := describeHsm(d, meta)
+
+	if hsm == nil {
+		log.Printf("[WARN] CloudHSMv2 HSM (%s) not found", d.Id())
+		d.SetId("")
+		return err
+	}
+
+	log.Printf("[INFO] Reading CloudHSMv2 HSM Information: %s", d.Id())
+
+	d.Set("cluster_id", hsm.ClusterId)
+	d.Set("subnet_id", hsm.SubnetId)
+	d.Set("availability_zone", hsm.AvailabilityZone)
+	d.Set("ip_address", hsm.EniIp)
+	d.Set("hsm_id", hsm.HsmId)
+	d.Set("hsm_state", hsm.State)
+	d.Set("hsm_eni_id", hsm.EniId)
+
+	return nil
+}
+
+func resourceAwsCloudHsm2HsmUpdate(d *schema.ResourceData, meta interface{}) error {
```

Comment:
If the resource does not support updates, this function should be removed and all `Optional`/`Required` attributes marked `ForceNew: true`